src/graph/visitor/PropertyTrackerVisitor.cpp

```diff
@@ -0,0 +1,223 @@
+/* Copyright (c) 2020 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License.
+ */
+
+#include "graph/visitor/PropertyTrackerVisitor.h"
+
+#include <sstream>
+#include <unordered_set>
+
+#include "common/expression/Expression.h"
+#include "graph/context/QueryContext.h"
+
+namespace nebula {
+namespace graph {
+
+PropertyTrackerVisitor::PropertyTrackerVisitor(const QueryContext *qctx,
+                                               GraphSpaceID space,
+                                               PropertyTracker &propsUsed,
+                                               const std::string &entityAlias)
+    : qctx_(qctx), space_(space), propsUsed_(propsUsed), entityAlias_(entityAlias) {
+  DCHECK(qctx != nullptr);
+}
+
+void PropertyTrackerVisitor::visit(TagPropertyExpression *expr) {
+  auto &tagName = expr->sym();
+  auto &propName = expr->prop();
+  auto ret = qctx_->schemaMng()->toTagID(space_, tagName);
+  if (!ret.ok()) {
+    status_ = std::move(ret).status();
+    return;
+  }
+  auto tagId = ret.value();
+  propsUsed_.vertexPropsMap[entityAlias_][tagId].emplace(propName);
+}
+
+void PropertyTrackerVisitor::visit(EdgePropertyExpression *expr) {
+  auto &edgeName = expr->sym();
+  auto &propName = expr->prop();
+  auto ret = qctx_->schemaMng()->toEdgeType(space_, edgeName);
+  if (!ret.ok()) {
+    status_ = std::move(ret).status();
+    return;
+  }
+  auto edgeType = ret.value();
+  propsUsed_.edgePropsMap[entityAlias_][edgeType].emplace(propName);
+}
+
+void PropertyTrackerVisitor::visit(LabelTagPropertyExpression *expr) {
+  auto status = qctx_->schemaMng()->toTagID(space_, expr->sym());
+  if (!status.ok()) {
+    status_ = std::move(status).status();
+    return;
+  }
+  auto &nodeAlias = static_cast<VariablePropertyExpression *>(expr->label())->prop();
+  auto &tagName = expr->sym();
+  auto &propName = expr->prop();
+  auto ret = qctx_->schemaMng()->toTagID(space_, tagName);
+  if (!ret.ok()) {
+    status_ = std::move(ret).status();
+    return;
+  }
+  auto tagId = ret.value();
+  propsUsed_.vertexPropsMap[nodeAlias][tagId].emplace(propName);
+}
+
+void PropertyTrackerVisitor::visit(InputPropertyExpression *expr) {
+  auto &colName = expr->prop();
+  propsUsed_.colsSet.emplace(colName);
+}
+
+void PropertyTrackerVisitor::visit(VariablePropertyExpression *expr) {
+  auto &colName = expr->prop();
+  propsUsed_.colsSet.emplace(colName);
+}
+
+// void PropertyTrackerVisitor::visit(AttributeExpression *expr) {
```

Comment:
why disable AttributeExpression

Reply:
Please see the TODO of the desc of this PR:
`At present, this PR only prunes vertex attributes, when the PR was almost reviewed, I can add the pruner of edge attribute.`